addons/purchase_stock/models/stock.py

```diff
@@ -38,21 +38,7 @@ def _get_price_unit(self):
         """ Returns the unit price for the move"""
         self.ensure_one()
         if self.purchase_line_id and self.product_id.id == self.purchase_line_id.product_id.id:
-            line = self.purchase_line_id
-            order = line.order_id
-            price_unit = line.price_unit
-            if line.taxes_id:
-                price_unit = line.taxes_id.with_context(round=False).compute_all(price_unit, currency=line.order_id.currency_id, quantity=1.0)['total_excluded']
-            if line.product_uom.id != line.product_id.uom_id.id:
-                price_unit *= line.product_uom.factor / line.product_id.uom_id.factor
-            if order.currency_id != order.company_id.currency_id:
-                # The date must be today, and not the date of the move since the move move is still
-                # in assigned state. However, the move date is the scheduled date until move is
-                # done, then date of actual move processing. See:
-                # https://github.com/odoo/odoo/blob/2f789b6863407e63f90b3a2d4cc3be09815f7002/addons/stock/models/stock_move.py#L36
```

Comment:
@sle-odoo this can be the only limitating thing for unifying the method, as the currency evaluation date is not the same. We have following options:

- Use `date_order` in both. This is current PR state. Not sure the business logic is right in this case.
- Let both methods. Then the call `_prepare_compute_all_values` should be called in both.
- Add a keyword parameter for passing an optional date for using in currency conversion.

Reply:
so

`_get_stock_move_price_unit` returns the price unit to set on the move when creating it (not yet validated)
`_get_price_unit returns` the price to be used when validating the move and create the according valuation entries

in the second case, we use the currency rate at the date of the validation and not of the move's creation (explanations [here](38eb5346c75fb97f215d59d208978cbc27a82b6e))

si i would go with your third options, add a `conversion_rate_date` kwarg to `_get_stock_move_price_unit` defaulting to None, and if not explicitely passed then the code should use `fields.Date.context_today(self)`

Reply:
Third method used then